alg/gdal_interpolateatpoint.cpp

```diff
@@ -34,23 +34,40 @@
 #include "gdalresamplingkernels.h"
 
 #include <algorithm>
+#include <complex>
 
-static bool
-GDALInterpExtractDEMWindow(GDALRasterBand *pBand,
-                           std::unique_ptr<DoublePointsCache> &cacheDEM, int nX,
-                           int nY, int nWidth, int nHeight, double *padfOut)
+template <typename T> bool areEqualReal(double dfNoDataValue, T dfOut);
+
+template <> bool areEqualReal(double dfNoDataValue, double dfOut)
+{
+    return ARE_REAL_EQUAL(dfNoDataValue, dfOut);
+}
+
+template <> bool areEqualReal(double dfNoDataValue, std::complex<double> dfOut)
+{
+    return ARE_REAL_EQUAL(dfNoDataValue, dfOut.real());
+}
+
+template <typename T>
```

Comment:
given the "double *padfAsDouble = reinterpret_cast<double *>(padfOut)", it would be prudent to at least add a comment that T should be double or ``std::complex<double>`` (perhaps some std::enable_if_t magic or static_assert could enforce that, but a simple code comment would do)

Reply:
Added a comment. In addition, `areEqualReal` should fail because it is not defined for other types.